operator/doublewrite/metricreporter.go

```diff
@@ -0,0 +1,159 @@
+	logger logrus.FieldLogger
+
+	interval time.Duration
+
+	kvStoreBackend            allocator.Backend
+	clientset                 k8sClient.Clientset
+	crdBackend                allocator.Backend
+	crdBackendWatcherStopChan chan struct{}
+
+	mgr *controller.Manager
+
+	metrics *Metrics
+}
+
+func registerDoubleWriteMetricReporter(p params) {
+	doubleWriteMetricReporter := &DoubleWriteMetricReporter{
+		logger:    p.Logger,
+		interval:  p.Cfg.Interval,
+		clientset: p.Clientset,
+		metrics:   p.Metrics,
+	}
+	p.Lifecycle.Append(doubleWriteMetricReporter)
+}
+
+func (g *DoubleWriteMetricReporter) Start(ctx cell.HookContext) error {
+	if option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadKVstore && option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadCRD {
+		return nil
+	}
+	g.logger.Info("Starting the Double Write Metric Reporter")
+
+	kvStoreBackend, err := kvstoreallocator.NewKVStoreBackend(kvstoreallocator.KVStoreBackendConfiguration{BasePath: cache.IdentitiesPath, Suffix: "", Typ: nil, Backend: kvstore.Client()})
+	if err != nil {
+		g.logger.WithError(err).Error("Unable to initialize kvstore backend for the Double Write Metric Reporter")
+		return err
+	}
+	g.kvStoreBackend = kvStoreBackend
+
+	crdBackend, err := identitybackend.NewCRDBackend(identitybackend.CRDBackendConfiguration{Store: nil, Client: g.clientset, KeyFunc: (&key.GlobalIdentity{}).PutKeyFromMap})
+	if err != nil {
+		g.logger.WithError(err).Error("Unable to initialize CRD backend for the Double Write Metric Reporter")
+		return err
+	}
+	g.crdBackend = crdBackend
+	// Initialize the CRD backend store
+	g.crdBackendWatcherStopChan = make(chan struct{})
+	go g.crdBackend.ListAndWatch(context.Background(), doublewrite.NoOpHandler{}, g.crdBackendWatcherStopChan)
+
+	g.mgr = controller.NewManager()
+	g.mgr.UpdateController("double-write-metric-reporter",
+		controller.ControllerParams{
+			Group:       controller.NewGroup("double-write-metric-reporter"),
+			RunInterval: g.interval,
+			DoFunc:      g.compareCRDAndKVStoreIdentities,
+		})
+
+	return nil
+}
+
+func (g *DoubleWriteMetricReporter) Stop(ctx cell.HookContext) error {
+	if g.mgr != nil {
+		g.mgr.RemoveAllAndWait()
+	}
+	close(g.crdBackendWatcherStopChan)
+	return nil
+}
+
+// difference counts the elements in `a` that aren't in `b` and returns a sample of differing elements (up to `maxElements`)
+func difference(a, b []idpool.ID, maxElements int) (int, []idpool.ID) {
+	mb := make(map[idpool.ID]struct{}, len(b))
+	for _, x := range b {
+		mb[x] = struct{}{}
+	}
+	c := 0
+	var diff []idpool.ID
+	for _, x := range a {
+		if _, found := mb[x]; !found {
+			c++
+			if len(diff) < maxElements {
+				diff = append(diff, x)
+			}
+		}
+	}
+	return c, diff
+}
+
+func (g *DoubleWriteMetricReporter) compareCRDAndKVStoreIdentities(ctx context.Context) error {
+	// Get CRD identities
+	crdIdentityIds, err := g.crdBackend.ListIDs(ctx)
```

Comment:
I'd wait for synchronization to complete before retrieving the data from the store, as it would otherwise report unnecessary spikes of mismatches until identities have been synchronized from k8s.